encoder/flatfile.go

```diff
@@ -0,0 +1,225 @@
+package encoder
+
+import (
+	"bufio"
+	"bytes"
+	"encoding/json"
+	"fmt"
+	common_datalayer "github.com/mimiro-io/common-datalayer"
+	"io"
+	"strings"
+)
+
+func NewFlatFileItemFactory() ItemFactory {
+	return &FlatFileItemFactory{}
+}
+
+type FlatFileItemFactory struct{}
+
+func (c *FlatFileItemFactory) NewItem() common_datalayer.Item {
+	return &FlatFileItem{data: make(map[string]any)}
+}
+
+type FlatFileItemWriter struct {
+	writer    io.WriteCloser
+	batchInfo *common_datalayer.BatchInfo
+	fields    []FlatFileField
+}
+
+// use position of field in the list to order fieldOrder
+type FlatFileField struct {
+	Name   string `json:"name"`
+	Length int    `json:"length"`
+	Ignore bool   `json:"ignore"`
+	Type   string `json:"type"`
+}
+
+func NewFlatFileItemWriter(sourceConfig map[string]any, data io.WriteCloser, batchInfo *common_datalayer.BatchInfo) (*FlatFileItemWriter, error) {
+	writer := &FlatFileItemWriter{writer: data, batchInfo: batchInfo}
+	// do this in function or here?
+	itemWriter, err2 := NewFlatFileWriteConfig(sourceConfig["fields"].([]map[string]interface{}), writer)
```

Comment:
This few lines look a bit off. It would look cleaner if the NewFlatFileWriteConfig returned the config item and then this was assigned into the writer, rather than the side effecting way it is done here. Also, I suspect it works but then the fact that the original writer is returned looks stranges, but also that if there is an error we still returned a writer, when it should probably be nil instead. As another relfection, at the moment there is only one property of the config the Fields property. This makes future extension to add new properties a bit trickier. I suggest having a FlatFileConfig struct with Fields being one property of that. Then instead of fields []FlatFileField on the FlatFileItemWriter there is a config *FlatFileConfig instead. The constructor function for the config takes the sourceConfig.

Reply:
Fixed, see changes